### PR TITLE
User specific enclave public key configuration in auth file

### PR DESCRIPTION
**Pull request by jframe**
## PR description
Allow a users enclave public key to be specified when creating the authentication configuration and create a JWT containing the enclave public key.
The users enclave public key on the JWT token will be needed in the future to support multi-tenancy for private transactions.

### CODE REVIEW

File: ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/TomlUserTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    assertThat(principal.getJsonArray("groups")).isEqualTo(new JsonArray(list("admin")));
+    assertThat(principal.getJsonArray("permissions"))
+        .isEqualTo(new JsonArray(list("eth:*", "perm:*")));
+    assertThat(principal.getJsonArray("roles")).isEqualTo(new JsonArray(list("net")));
+    assertThat(principal.getString("enclavePublicKey"))
+        .isEqualTo("A1aVtMxLCUHmBVHXoZzzBgPbW/wj5axDpW9X8l91SGo=");
+  }
+
+  @Test
+  public void createsPrincipleWithOptionalValues() {
```
Comment (macfarla) [inline ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/TomlUserTest.java]: naming - is this more like createsPrincipleWithOnlyMandatoryValues

Comment (macfarla) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/TomlUserTest.java]: or ...OnlyRequiredValues

Comment (jframe) [reply ethereum/api/src/test/java/org/hyperledger/besu/ethereum/api/jsonrpc/authentication/TomlUserTest.java]: That's a better name. Renamed to createsPrincipleWithOnlyRequiredValues

File: besu/src/test/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidatorTest.java [hunk 2]
```diff
@@ -27,11 +27,16 @@
...
 
-  private static final String CORRECT_TOML = "/auth_correct.toml";
-  private static final String DUPLICATE_USER_TOML = "/auth_duplicate_user.toml";
-  private static final String INVALID_TOML = "/auth_invalid.toml";
-  private static final String INVALID_VALUE_TOML = "/auth_invalid_value.toml";
-  private static final String NO_PASSWORD_TOML = "/auth_no_password.toml";
+  private static final String CORRECT_TOML = "/rpcauth/auth_correct.toml";
+  private static final String DUPLICATE_USER_TOML = "/rpcauth/auth_duplicate_user.toml";
+  private static final String INVALID_TOML = "/rpcauth/auth_invalid.toml";
+  private static final String INVALID_GROUPS_VALUE_TOML = "/rpcauth/auth_invalid_groups_value.toml";
```
Comment (rain-on) [inline besu/src/test/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidatorTest.java]: do we want to put in the groups etc. at this stage? Or just the PRIVACY_KEY?

File: besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java [hunk 3]
```diff
@@ -97,7 +99,20 @@ private static boolean verifyAllEntriesHaveValues(final TomlParseResult tomlPars
...
+    if (key.endsWith(PRIVACY_PUBLIC_KEY)) {
+      return verifyString(key, tomlParseResult);
+    } else {
+      return verifyArray(key, tomlParseResult);
+    }
+  }
+
+  private static boolean verifyString(final String key, final TomlParseResult tomlParseResult) {
+    final String value = tomlParseResult.getString(key);
+    return tomlParseResult.isString(key) && value != null && !value.isEmpty();
```
Comment (CjHare) [inline besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java]: Based on the source code of `isString()` you'll want to perform the null check before invoking `tomlParseResult.isString(key)`

File: besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java [hunk 4]
```diff
@@ -97,7 +99,20 @@ private static boolean verifyAllEntriesHaveValues(final TomlParseResult tomlPars
...
+
+  private static boolean verifyEntry(final TomlParseResult tomlParseResult, final String key) {
+    if (key.endsWith(PRIVACY_PUBLIC_KEY)) {
+      return verifyString(key, tomlParseResult);
+    } else {
+      return verifyArray(key, tomlParseResult);
+    }
+  }
+
+  private static boolean verifyString(final String key, final TomlParseResult tomlParseResult) {
```
Comment (CjHare) [inline besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java]: nit; is there a reason the ordering of the input parameters is inconsistent between `verifyEntry` and `verifyString`?

File: besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java (hunk 3 above)
Comment (CjHare) [reply besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java]: The reason is the source version of that method for me comes from `TomlTable`, where there is the require not null check 😉 
```  
default boolean isString(String dottedKey) {
    requireNonNull(dottedKey);
    return isString(parseDottedKey(dottedKey));
  }
```

Comment (CjHare) [reply besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java]: Nevermind, my bad, I misread the `key` as `value` 🙄

File: besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java (hunk 4 above)
Comment (jframe) [reply besu/src/main/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidator.java]: nope, nice catch.
done

File: besu/src/test/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidatorTest.java (hunk 2 above)
Comment (jframe) [reply besu/src/test/java/org/hyperledger/besu/cli/custom/RpcAuthFileValidatorTest.java]: This is just testing the parsing of groups that already exists in Auth file parsing. It doesn't add groups or roles as claims.

**Pull request merged**